File: backend/app/api/courses.py
from fastapi import APIRouter
from fastapi.responses import JSONResponse

from app.schemas.course import (
    CourseCategory,
    CourseClassType,
    CourseDetailResponse,
    CourseListResponse,
    CourseStatus,
)
from app.services import course_service

router = APIRouter(prefix="/courses", tags=["courses"])


@router.get("", response_model=CourseListResponse)
def get_courses(
    status: CourseStatus | None = None,
    classType: CourseClassType | None = None,
    category: CourseCategory | None = None,
    search: str | None = None,
) -> CourseListResponse:
    courses = course_service.search_courses(
        status=status, class_type=classType, category=category, search=search
    )
    return CourseListResponse(courses=courses)


@router.get("/{course_id}")
def get_course(course_id: str):
    course = course_service.get_course_by_id(course_id)
    if course is None:
        return JSONResponse(
            status_code=404,
            content={
                "error": {
                    "code": "COURSE_NOT_FOUND",
                    "message": "과목을 찾을 수 없습니다.",
                }
            },
        )
    return CourseDetailResponse(course=course)

File: frontend/src/lib/api/courses.ts
import type { Course, CourseClassType, CourseCategory, CourseStatus } from "@/types";
import { mockCourses } from "./store";

export interface GetCoursesParams {
  status?: CourseStatus;
  classType?: CourseClassType;
  category?: CourseCategory;
  search?: string;
}

// GET /courses
export async function getCourses(params: GetCoursesParams = {}): Promise<{ courses: Course[] }> {
  let courses = mockCourses;
  if (params.status) courses = courses.filter((c) => c.status === params.status);
  if (params.classType) courses = courses.filter((c) => c.classType === params.classType);
  if (params.category) courses = courses.filter((c) => c.category === params.category);
  if (params.search) {
    const q = params.search.toLowerCase();
    courses = courses.filter(
      (c) => c.name.toLowerCase().includes(q) || c.professor.toLowerCase().includes(q)
    );
  }
  return { courses };
}

// GET /courses/{courseId}
export async function getCourse(courseId: string): Promise<{ course: Course | null }> {
  const course = mockCourses.find((c) => c.id === courseId) ?? null;
  return { course };
}
